pkg/endpoint/policy.go

```diff
@@ -80,139 +69,49 @@ func getSecurityIdentities(labelsMap *identityPkg.IdentityCache, selector *api.E
-// Returns the number of errors that occurred while when applying the
-// policy.
-// Applies for L3-dependent L4, and not for L4-only policy.
-func (e *Endpoint) applyNewFilter(identities *identityPkg.IdentityCache,
-	filter *policy.L4Filter, direction policymap.TrafficDirection) int {
+func (e *Endpoint) convertL4FilterToPolicyMapKeys(identities *identityPkg.IdentityCache,
+	filter *policy.L4Filter, direction policymap.TrafficDirection) []policymap.PolicyKey {
 
+	keysToAdd := []policymap.PolicyKey{}
 	port := uint16(filter.Port)
 	proto := uint8(filter.U8Proto)
 
-	errors := 0
 	for _, sel := range filter.Endpoints {
 		for _, id := range getSecurityIdentities(identities, &sel) {
 			srcID := id.Uint32()
-			if e.PolicyMap.Exists(srcID, port, filter.U8Proto, direction) {
-				e.getLogger().WithField("l4Filter", filter).Debug("L4 filter exists")
-				continue
-			}
-			if err := e.PolicyMap.Allow(srcID, port, filter.U8Proto, direction); err != nil {
-				e.getLogger().WithFields(logrus.Fields{
-					logfields.PolicyID: srcID,
-					logfields.Port:     port,
-					logfields.Protocol: proto}).WithError(err).Warn(
-					"Update of l4 policy map failed")
-				errors++
+			keyToAdd := policymap.PolicyKey{
+				Identity:         srcID,
+				DestPort:         byteorder.HostToNetwork(port).(uint16),
+				Nexthdr:          proto,
+				TrafficDirection: direction.Uint8(),
 			}
+			keysToAdd = append(keysToAdd, keyToAdd)
 		}
 	}
-	return errors
+	return keysToAdd
 }
 
-// Looks for mismatches between 'oldPolicy' and 'newPolicy', and fixes up
-// this Endpoint's BPF PolicyMap to reflect the new L3+L4 combined policy.
-// Returns a map that represents all L3-dependent L4 PolicyMap entries that were attempted
-// to be added;
-// and a map that represents all L3-dependent L4 PolicyMap entries that were attempted
-// to be removed;
-// it maps to whether they were removed successfully (true or false)
-// TODO (it maps to rule contexts); 'whether they were removed successfully' doesn't
-// make sense - is this what L4Installed means?
-func (e *Endpoint) applyL4PolicyLocked(oldIdentities, newIdentities *identityPkg.IdentityCache,
-	oldL4Policy, newL4Policy *policy.L4Policy) (err error) {
-
-	var errors = 0
-
-	// Need to iterate through old L3-L4 policy and remove all PolicyMap entries
-	// for both ingress and egress.
-	if oldL4Policy != nil {
-		for _, filter := range oldL4Policy.Ingress {
-			e.removeOldFilter(oldIdentities, &filter, policymap.Ingress)
-		}
-
-		for _, filter := range oldL4Policy.Egress {
-			e.removeOldFilter(oldIdentities, &filter, policymap.Egress)
-		}
-		errors += e.sweepFilters(oldL4Policy, newIdentities)
-	}
-
-	// No new entries to add to PolicyMap, so simply return.
-	if newL4Policy == nil {
-		return nil
-	}
+func (e *Endpoint) computeDesiredL4PolicyMapEntries(labelsMap *identityPkg.IdentityCache, desiredL4Policy *policy.L4Policy) (keysToAdd map[policymap.PolicyKey]struct{}) {
```

Comment:
`IdentityCache` is a map, so it already has reference semantics. Don't pass it as a pointer.

Reply:
Fixed.